packages/jest-haste-map/src/index.js

```diff
@@ -299,30 +301,66 @@ class HasteMap extends EventEmitter {
     mocks: Object,
     filePath: Path,
     workerOptions: ?{forceInBand: boolean},
-  ): ?Promise<void> {
-    const setModule = (id: string, module: ModuleMetaData) => {
+  ): ?Promise<?{id: string, candidates: Array<ModuleMetaData>}> {
+
+    const updateModuleFromCandidates = (
+      id: string,
+      platform: string,
+      candidates: Array<ModuleMetaData>,
+    ) => {
       if (!map[id]) {
         map[id] = Object.create(null);
       }
-      const moduleMap = map[id];
-      const platform = getPlatformExtension(module[H.PATH]) ||
-        H.GENERIC_PLATFORM;
-      const existingModule = moduleMap[platform];
-      if (existingModule && existingModule[H.PATH] !== module[H.PATH]) {
-        const message = `jest-haste-map: @providesModule naming collision:\n` +
-          `  Duplicate module name: ${id}\n` +
-          `  Paths: ${module[H.PATH]} collides with ` +
-          `${existingModule[H.PATH]}\n\nThis ` +
-          `${this._options.throwOnModuleCollision ? 'error' : 'warning'} ` +
-          `is caused by a @providesModule declaration ` +
-          `with the same name across two different files.`;
-        if (this._options.throwOnModuleCollision) {
-          throw new Error(message);
+      const modulesByPlatform = map[id];
+      if (candidates.length === 1) {
+        modulesByPlatform[platform] = candidates[0];
+      } else {
+        delete modulesByPlatform[platform];
+      }
+      return {id, candidates};
+    }
+
+    const setModule = (id: string, module: ModuleMetaData) => {
+
+      const moduleFilePath = module[H.PATH];
+      const platform =
+        getPlatformExtension(moduleFilePath) || H.GENERIC_PLATFORM;
+
+      const existingID = this._moduleIDsByFilePath.get(moduleFilePath);
+      if (existingID != null) {
+        const moduleArraysByPlatform = this._moduleArraysByIDAndPlatform.get(existingID);
+        if (moduleArraysByPlatform == null) {
+          throw new Error('could not find data for existing ID');
```

Comment:
Are these just invariants that will never happen or do you expect user code to run into this for some reason? If it is user code, then please prefix them with `jest-haste-map: ` and if it is internal code, please prefix with `invariant: `. Regardless of which it is, please capitalize the first word and add a `.` at the end of the sentence :)